opentasks-provider/src/main/java/org/dmfs/provider/tasks/processors/instances/Validating.java

```diff
@@ -42,51 +75,110 @@ public Validating(EntityProcessor<InstanceAdapter> delegate)
+                            TaskContract.Instances.TASK_ID,
+                            TaskContract.Instances.ORIGINAL_INSTANCE_ID,
+                            TaskContract.Instances.INSTANCE_ORIGINAL_TIME,
+                            TaskContract.Instances.INSTANCE_START,
+                            TaskContract.Instances.INSTANCE_DUE),
+                    new String[] {
+                            instanceId.value().toString(),
+                            instanceId.value().toString(),
+                            timeStampString,
+                            timeStampString,
+                            timeStampString },
+                    null,
+                    null,
+                    null))
+            {
+                if (c.getCount() > 0)
+                {
+                    throw new IllegalArgumentException(String.format(Locale.ENGLISH, "Instance %s of task %d already exists",
+                            entityAdapter.taskAdapter().valueOf(TaskAdapter.ORIGINAL_INSTANCE_TIME).toString(), instanceId.value()));
+                }
+            }
         }
+    }
+
 
+    private void validateValues(InstanceAdapter instanceAdapter)
+    {
         // actually, no instance value can be changed, the instance table only allows for updating task values
-        if (instanceAdapter.isUpdated(InstanceAdapter._ID) ||
-                instanceAdapter.isUpdated(InstanceAdapter.INSTANCE_START) ||
-                instanceAdapter.isUpdated(InstanceAdapter.INSTANCE_START_SORTING) ||
-                instanceAdapter.isUpdated(InstanceAdapter.INSTANCE_DUE) ||
-                instanceAdapter.isUpdated(InstanceAdapter.INSTANCE_DUE_SORTING) ||
-                instanceAdapter.isUpdated(InstanceAdapter.INSTANCE_ORIGINAL_TIME) ||
-                instanceAdapter.isUpdated(InstanceAdapter.DISTANCE_FROM_CURRENT) ||
-                instanceAdapter.isUpdated(InstanceAdapter.TASK_ID))
+        if (new First<>(new Sieved<>(instanceAdapter::isUpdated, INSTANCE_FIELD_ADAPTERS)).isPresent())
         {
             throw new IllegalArgumentException("Instance columns are read-only.");
         }
 
         TaskAdapter taskAdapter = instanceAdapter.taskAdapter();
         // By definition, single instances don't have a recurrence set on their own, hence changes to the recurrence fields are not allowed.
-        if (taskAdapter.isUpdated(TaskAdapter.RRULE) ||
-                taskAdapter.isUpdated(TaskAdapter.RDATE) ||
-                taskAdapter.isUpdated(TaskAdapter.EXDATE))
+        if (new First<>(new Sieved<>(taskAdapter::isUpdated, RECURRENCE_FIELD_ADAPTERS)).isPresent())
+        {
+            throw new IllegalArgumentException("Recurrence values can not be modified through the instances table.");
+        }
+    }
+
+
+    private void validateOriginalInstanceValues(InstanceAdapter instanceAdapter)
```

Comment:
Is there a test case for this already?
I suppose all fields would need to be tested individually, for insert and update, so that may need some re-usable solution. And I also know that more tests are probably coming in the coming PRs, but I am a bit concerned that it's easy to loose track what to get back to. So perhaps start adding TODOs for test-cases in the test class?

Reply:
done #626